docs/standard/native-interop/tutorial-comwrappers.md

```diff
@@ -0,0 +1,936 @@
+        StoreString(inst, len, str);
+    }
+}
+```
+
+There are two important things to note in this example. The first is `DynamicInterfaceCastableImplementationAttribute`. This attribute is required on any type that is returned from a `IDynamicInterfaceCastable` method. It has the added benefit of making IL Linking easier which means NativeAOT scenarios are more reliable. The second is the cast to `DemoNativeDynamicWrapper`. This is part of the dynamic nature of `IDynamicInterfaceCastable`. The type that is returned from `IDynamicInterfaceCastable.GetInterfaceImplementation()` is used to blanket the type that implements `IDynamicInterfaceCastable`.
+
+Regardless of which Native Object Wrapper is used, the ability to invoke functions on a COM instance is needed. The implementation of `IDemoStoreTypeNativeWrapper.StoreString()` can serve as an example employing `unmanaged` C# function pointers.
+
+```csharp
+public static void StoreString(IntPtr inst, int len, string? str)
+{
+    IntPtr strLocal = Marshal.StringToCoTaskMemUni(str);
+    int hr = ((delegate* unmanaged<IntPtr, int, IntPtr, int>)(*(*(void***)inst + 3 /* IDemoStoreType.StoreString slot */)))(inst, len, strLocal);
+    if (hr != 0)
+    {
+        Marshal.FreeCoTaskMem(strLocal);
+        Marshal.ThrowExceptionForHR(hr);
+    }
+}
+```
+
+The dereferencing of the COM instance to access its vtable implementation warrants examining. The COM ABI defines that the first pointer of an object is to the type's vtable and from there the desired slot can be accessed. Let's assume the address of the COM object is `0x10000`. The first pointer sized value should be the address of the vtable &ndash; in this example `0x20000`. Once we are at the vtable we look for the 4th slot (3 in zero-based indexing) to access the `StoreString()` implementation.
+
+```
+COM instance
+0x10000  0x20000
+
+VTable for IDemoStoreType
+0x20000  <Address of QueryInterface>
+0x20008  <Address of AddRef>
+0x20010  <Address of Release>
+0x20018  <Address of StoreString>
+```
+
+Having the function pointer then allows us to dispatch to that member function on that object by passing the object instance as the fist parameter &ndash; that should look familiar based on the function definitions when implementing the Managed Object Wrapper.
+
+Once the `CreateObject()` method is implemented the `ComWrappers` subclass will be able to produce Native Object Wrappers for COM instances that implement both `IDemoGetType` and `IDemoStoreType`.
+
+```csharp
+IntPtr iunk = ...; // Get a COM instance from native code.
+object rcw = cw.GetOrCreateObjectForComInstance(ccw, CreateObjectFlags.UniqueInstance);
+```
+
+### Step 4 &ndash; Native Object Wrapper lifetime details
+
+The `ComputeVtables()` and `CreateObject()` implementations discussed some wrapper lifetime details but further details need discussion. This can be a short step but can also significantly increase the complexity of the `ComWrappers` design.
```

Comment:
```suggestion
The `ComputeVtables()` and `CreateObject()` implementations covered some wrapper lifetime details, but there are further considerations. While this can be a short step, it can also significantly increase the complexity of the `ComWrappers` design.
```